openedx_tagging/core/tagging/models.py

```diff
@@ -456,3 +463,103 @@ def resync(self) -> bool:
+
+        Subclasses should override this method to perform any additional validation for the particular type of object tag.
+        """
+        # Must have a valid object id/type:
+        return self.object_id and self.object_type
+
+    def is_valid(self, check_taxonomy=True, check_tag=True, check_object=True) -> bool:
+        """
+        Returns True if this ObjectTag is valid for use with a free-text taxonomy.
+
+        Subclasses should override this method to perform any additional validation for the particular type of object tag.
+
+        If `check_taxonomy` is False, then we skip validating the object tag's taxonomy reference.
+        If `check_tag` is False, then we skip validating the object tag's tag reference.
+        If `check_object` is False, then we skip validating the object ID/type.
+        """
+        # Must be linked to a free-text taxonomy
+        if check_taxonomy and (
+            not self.taxonomy_id or not self.taxonomy.allow_free_text
+        ):
+            return False
+
+        # Open taxonomies don't need an associated tag, but we need a value.
+        if check_tag and not self.value:
+            return False
+
+        if check_object and not self._check_object():
+            return False
+
+        return True
+
+
+class ClosedObjectTag(OpenObjectTag):
+    """
+    Object tags linked to a closed taxonomy, where the available tag value options are known.
+    """
+
+    class Meta:
+        proxy = True
+
+    @classmethod
+    def valid_for(cls, taxonomy: Taxonomy = None, tag: Tag = None, **kwargs) -> bool:
+        """
+        Returns True if it's a closed taxonomy and there's a tag
+        """
+        return tag and taxonomy and not taxonomy.allow_free_text
+
+    def is_valid(self, check_taxonomy=True, check_tag=True, check_object=True) -> bool:
+        """
+        Returns True if this ObjectTag is valid for use with a closed taxonomy.
+
+        Subclasses should override this method to perform any additional validation for the particular type of object tag.
+
+        If `check_taxonomy` is False, then we skip validating the object tag's taxonomy reference.
+        If `check_tag` is False, then we skip validating the object tag's tag reference.
+        If `check_object` is False, then we skip validating the object ID/type.
+        """
+        # Must be linked to a closed taxonomy
+        if check_taxonomy and (not self.taxonomy_id or self.taxonomy.allow_free_text):
+            return False
+
+        # Closed taxonomies require a Tag
+        if check_tag and not self.tag_id:
+            return False
+
+        if check_tag and check_taxonomy and (self.tag.taxonomy != self.taxonomy):
+            return False
+
+        if check_object and not self._check_object():
+            return False
+
+        return True
+
+
+# Register the object tag classes in reverse order for how we want them considered
+register_object_tag_class(ObjectTag)
```

Comment:
@pomegranited Maybe we can explicitly add the order in this function? I say this in case object tag classes are registered in other files/modules, in those cases I would not be sure which one goes first or which one goes second

Reply:
Sure thing, done with [f260ddd](https://github.com/openedx/openedx-learning/pull/62/commits/f260ddd8bf78840329f2ccff8862f0d9a60126e3)